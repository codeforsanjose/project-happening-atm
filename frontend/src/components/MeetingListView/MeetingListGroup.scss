@import "../../utils/_colors.scss";

.MeetingListGroup {
  .group-header {
    margin-bottom: 0.8em;

    .group-button {
      padding: 1em;
      background-color: transparent;
      color: $cityBlue;
      border-style: solid;
      border-width: 0;
      border-top-width: 1px;
      border-color: $gray;
      font-weight: bold;
      cursor: pointer;

      .group-main-text:before {
        display: inline-block;
        content: '';
        height: 0;
        width: 0;
        margin-right: 1em;
        border-top: 0.5em solid transparent;
        border-bottom: 0.5em solid transparent;
        border-left: 0.5em solid currentColor;
        transition: 0.1s ease-in transform;
      }

      &[aria-expanded='true'] .group-main-text::before,
      &[aria-selected='true'] .group-main-text::before {
          transform: rotate(90deg);
      }

      .button-text {
        display: flex;
        justify-content: space-between;

        @media only screen and (min-device-width: 812px) {
          justify-content: left;

          .group-sub-text {
            margin-left: 0.8em;
          }
        }
      }
    }
  }

  .group-items {
    animation: fadein 0.3s ease-in;

    @keyframes fadein {
      0% {
        opacity: 0;
      }

      100% {
        opacity: 1;
      }
    }
  }
}