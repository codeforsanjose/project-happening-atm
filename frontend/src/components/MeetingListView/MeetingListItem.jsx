import React from 'react';
import { useTranslation } from 'react-i18next';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { toDateString, toTimeString } from '../../utils/timestampHelper';
import './MeetingListItem.scss';
import { StatusInProgress } from '../../utils/_icons';
import isAdmin from '../../utils/isAdmin';

// Component imports
import {
  AdminMeetingItemLinks,
  PastMeetingItemLinks,
  PendingMeetingItemLinks,
} from './MeetingListItemLinks';

/**
 * A meeting displayed as an accordion sub item.
 *
 * props:
 *    item
 *      Object that represents a meeting item.
 *      {
 *        id: Number id of item
 *        meeting_start_timestamp: String starting timestamp (Unix time)
 *        status: String status of item - ['PENDING', 'IN PROGRESS', 'CLOSED']
 *      }
 */

function MeetingListItem({ item }) {
  const { t } = useTranslation();
  // eslint-disable-next-line camelcase
  const { id, meeting_start_timestamp, status } = item;
  const date = toDateString(meeting_start_timestamp, 'dddd, MMM D');
  const time = toTimeString(meeting_start_timestamp);
  const isInProgress = status === 'IN PROGRESS';
  const isEnded = status === 'ENDED';

  const isCurrentUserAdmin = isAdmin();
  const PublicLinks =
    status === 'CLOSED' ? PastMeetingItemLinks : PendingMeetingItemLinks;
  const MeetingItemLinks = isCurrentUserAdmin
    ? AdminMeetingItemLinks
    : PublicLinks;

  return (
    <div
      className={classnames('MeetingListItem', { 'in-progress': isInProgress })}
    >
      <Link to={`meeting/${id}`} className="meeting-date">
        <h3>{date}</h3>
      </Link>
      <div
        className={classnames('meeting-status', {
          'progress-wrapper-started': isInProgress,
        })}
      >
        {isInProgress && (
          <>
            <StatusInProgress className="status-icon" />
            <span>{t('meeting.status.short.in-progress')}</span>
          </>
        )}
      </div>
      <Link to={`meeting/${id}`} className="meeting-time">
        <div>{time}</div>
      </Link>
      <MeetingItemLinks meeting={item} isInProgress={isInProgress} />
    </div>
  );
}

export default MeetingListItem;
