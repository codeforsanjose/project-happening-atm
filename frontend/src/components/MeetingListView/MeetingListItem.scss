@import "../../utils/_colors.scss";

.MeetingListItem {
  box-sizing: border-box;
  margin-bottom: 1em;
  padding: 1em;
  display: grid;
  grid-template-columns: 1fr;
  grid-gap: 0.5em;
  border: 2px solid $cloudy;
  border-radius: 5px;
  background-color: #fff;

  @media screen and (max-width: 500px) {
    padding: 0.8em;
  }

  @media only screen
  and (min-device-width: 812px) {
    grid-template-columns: 1.4em 2fr 1fr 1.4em;
    grid-template-rows: 1em 1fr 3em 1fr 1.4em;
    grid-gap: 0;
    padding: 0;

    h3 {
      padding-top: 1.4em;
      padding-left: 1.4em;
    }

    .meeting-date {
      grid-row: 1 / -1;
      grid-column: 1 / -1;
    }

    .meeting-time {
      grid-row: 2 / 5;
      grid-column: 2;
      font-size: 1.2em;
      padding-left: 2px;
      padding-bottom: 1em;

      display: flex;
      align-items: center;
    }

    .meeting-status {
      grid-row: 2;
      grid-column: 3;

      display: flex;
      align-items: center;
      justify-content: flex-end;
    }

    .progress-wrapper-started {
      color: $statusIndicator;
      font-weight: 800;
      font-size: 1.2rem;
      padding: 0.4em 0.7em;

      .status-icon {
        fill: $lightGreen;
        margin-right: 0.5em;
        position: relative;
        transform: scale(0.85);
      }
    }

    .meeting-links {
      grid-row: 4 / 5;
      grid-column: 2 / 3;
      justify-self: left;

      a {
        margin-right: 3em;
      }

      button {
        background-color: inherit;
        border: none;
        color: inherit;
        cursor: pointer;
      }
    }
  }

  a {
    color: inherit;
    text-decoration: none;

    .meeting-date {
      margin: 0;
      display: inline-block;
      font-weight: 600;
    }
  }

  h3 {
    margin: 0;
  }

  .meeting-links {
    display: flex;
    flex-direction: row;

    @media screen and (max-width: 500px) {
      margin-top: 1rem;
      align-items: center;
    }

    a {
      text-decoration: none;

      .upload-agenda-button, .delete-agenda-button {
        padding: 10px 15px;
        border-radius: 5px;
        border: transparent;

        @media screen and (max-width: 500px) {
          padding: 5px 8px;
        }

        svg {
          @media screen and (max-width: 500px) {
            width: 15px;
          }
        }
      }
      
      .upload-agenda-button {
        color: $cityBlue;
        fill: $cityBlue;
        background-color: $atmGold;
      }

      .delete-agenda-button {
        color: $errorRed;
        fill: $errorRed;
        background-color: transparent;
      }

      .link {
        display: flex;
        align-items: center;
        font-size: 0.9em;
        width: max-content;

        p {
          margin: 0;
          margin-left: 0.4em;

          @media screen and (max-width: 500px) {
            font-size: 12px;
          }
        }

        &:hover {
          text-decoration: underline;
        }
      }

    }
    
    .disabled-link {
      color: gray;
      pointer-events: none;
      
      svg {
        fill: gray;
      }
    }
    
  }

  &:hover {
    border: 2px solid $lightBlue;
  }

  &.in-progress {
    border: 2px solid $cityBlue;
    background-color: $itemInProgress;
  }
}
