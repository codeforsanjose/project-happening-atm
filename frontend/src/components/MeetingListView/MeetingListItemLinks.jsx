/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { useTranslation } from 'react-i18next';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import useCSVUpload from '../../hooks/useCSVUpload';
import useDeleteMeeting from '../../hooks/useDeleteMeeting';
import { isFutureTimestamp } from '../../utils/timestampHelper';
import '../../utils/generalStyles.scss';

// Asset imports
import {
  CalendarTodayIcon,
  ViewAgendaIcon,
  FormatListNumberedIcon,
  PlayWithCircleOutlineIcon,
  DeleteIcon,
} from '../../utils/_icons';

function AdminMeetingItemLinks({ meeting }) {
  const { t } = useTranslation();
  const [openUploadModal, UploadModal] = useCSVUpload(meeting);
  const [openDeleteModal, DeleteModal] = useDeleteMeeting(meeting);

  const isFutureItem = isFutureTimestamp(meeting.meeting_start_timestamp);
  const pastClasses = 'meeting-links visibility-hidden';
  const futureClasses = 'meeting-links';

  return (
    <div className={isFutureItem ? futureClasses : pastClasses}>
      <Link to="#">
        <button
          type="button"
          className="link upload-agenda-button"
          onClick={openUploadModal}
        >
          <ViewAgendaIcon />
          <p>{t('meeting.actions.upload-new-agenda')}</p>
        </button>
      </Link>
      <Link to="#">
        <button
          type="button"
          className="link delete-agenda-button"
          onClick={openDeleteModal}
        >
          <DeleteIcon />
          <p>{t('meeting.list.delete-meeting.button')}</p>
        </button>
      </Link>

      <UploadModal />
      <DeleteModal />
    </div>
  );
}

function PastMeetingItemLinks() {
  const { t } = useTranslation();

  return (
    <div className="meeting-links">
      <Link to="#">
        <div className="link">
          <FormatListNumberedIcon />
          <p>{t('meeting.list.minutes.button')}</p>
        </div>
      </Link>
      <Link to="#">
        <div className="link">
          <PlayWithCircleOutlineIcon />
          <p>{t('meeting.list.recording.button')}</p>
        </div>
      </Link>
    </div>
  );
}

function PendingMeetingItemLinks({ meeting, isInProgress }) {
  const { t } = useTranslation();

  return (
    <div className="meeting-links">
      <Link to="#" className={classnames({ 'disabled-link': isInProgress })}>
        <div className="link">
          <CalendarTodayIcon />
          <p>{t('meeting.list.calendar-export.button')}</p>
        </div>
      </Link>
      <Link to={`meeting/${meeting.id}`}>
        <div className="link">
          <ViewAgendaIcon />
          <p>{t('meeting.list.agenda.button')}</p>
        </div>
      </Link>
    </div>
  );
}

export { AdminMeetingItemLinks, PastMeetingItemLinks, PendingMeetingItemLinks };
