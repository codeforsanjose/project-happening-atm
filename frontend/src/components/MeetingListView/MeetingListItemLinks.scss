@import '../../utils/_colors.scss';

.meeting-links {
  display: flex;
  flex-direction: row;
  text-decoration: none;

  a {
    .upload-agenda-button,
    .delete-agenda-button {
      padding: 10px 15px;
      border-radius: 5px;
      border: transparent;
      cursor: pointer;

      @media screen and (max-width: 500px) {
        padding: 5px 8px;
      }

      svg {
        @media screen and (max-width: 500px) {
          width: 15px;
        }
      }
    }

    .upload-agenda-button {
      color: $cityBlue;
      fill: $cityBlue;
      background-color: $atmGold;
    }

    .delete-agenda-button {
      color: $errorRed;
      fill: $errorRed;
      background-color: transparent;
    }

    .link {
      display: flex;
      align-items: center;
      font-size: 0.9em;
      width: max-content;

      p {
        margin: 0;
        margin-left: 0.4em;

        @media screen and (max-width: 500px) {
          font-size: 13px;
        }
      }

      &:hover {
        text-decoration: underline;
      }
    }
  }

  .disabled-link {
    color: gray;
    pointer-events: none;

    svg {
      fill: gray;
    }
  }
}
