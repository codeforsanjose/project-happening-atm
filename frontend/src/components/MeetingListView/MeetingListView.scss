@import "../../utils/colors";

.MeetingListView {
  margin-top: 4em;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  font-family: "Poppins", sans-serif;

  .meeting-list-header {
    padding: 0.5em 1.25em;

    .logo {
      width: 77px;
    }

    h2 {
      margin: 0.5em 0;

      // &::after {
      //   display: block;
      //   content: "";
      //   margin-top: 0.2em;
      //   height: 2px;
      //   width: 8em;
      //   background-color: $cityBronze;
      // }
    }

    .sub-header {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center;
      margin: 0;
      font-size: 16px;
    }
    .meeting-links {
      a {
        color: $lightBlue;
        text-decoration: none;
      }
    }
    button {
      display: flex;
      align-items: center;
      padding: 8px 16px;
      border: none;
      border-radius: 0.5rem;
      background-color: $atmGold;
      color: $lightBlue;
      font-size: 14px;
      font-weight: 400;
      // cursor: pointer;

      p {
        margin: 0;
        margin-left: 0.4em;
      }
      &:hover {
        box-shadow: 0 0 4px $cityBronze inset;

      }
    }

    @media only screen and (min-width: 991px){
      margin: 0 8em;
      padding-left: 4.5em;

      .logo {
        display: none;
      }

      .sub-header {
        display: block;
      }

      h2 {
        font-size: 2em;
      }
    }
  }

  .meeting-list-content {
    padding: 1.25em 10px;
    flex-grow: 1;
    background-color: $backgroundWhite;

    .complete-toggle {
      padding: 0;
      margin-bottom: 1.25em;
      display: inline-flex;
      align-items: center;
      border: none;
      font-size: 1em;
      background-color: transparent;
      // cursor: pointer;

      p {
        margin: 0 0.5em;
      }
    }

    @media only screen and (min-width: 991px){
      padding: 1.25em 200px;
    }
    
  }

  .loader {
    margin: 10em auto;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .footerSearchLink {
    display: block;
    justify-content: center;
    text-align: center;
    font-size: 14px;
    line-height: 150%;
  }
  .footerSearchLink a:link {
    color: $cityBlue;
    font-weight: bold;
  }
}
