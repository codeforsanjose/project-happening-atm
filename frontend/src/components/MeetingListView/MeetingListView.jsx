import React, { useState, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import { useQuery } from '@apollo/client';
import { Accordion } from 'react-accessible-accordion';
import {
  groupMeetingsByDate,
  isFutureTimestamp,
} from '../../utils/timestampHelper';
import {
  GET_ALL_MEETINGS,
  GET_ALL_UPCOMING_MEETINGS,
} from '../../graphql/query';

import isAdmin from '../../utils/isAdmin';
import './MeetingListView.scss';

// Component imports
import NavBarHeader from '../NavBarHeader/NavBarHeader';
import { AdminMeetingListViewLinks } from './MeetingListViewLinks';
import MeetingListGroup from './MeetingListGroup';
import Spinner from '../Spinner/Spinner';
import PollIntervals from '../../constants/PollStatusIntervals';
// Asset imports
import MEETINGS_ARCHIVE_URL from '../../constants/MeetingsArchiveURL';
import cityLogo from '../../assets/SanJoseCityLogo.png';
import {
  CheckedCheckboxIcon,
  UncheckedCheckboxIcon,
  ExternalSite,
} from '../../utils/_icons';

/**
 * Main view component to display a list of upcoming/past meetings.
 *
 * state:
 *    navToggled
 *      Boolean value indicating if the header nav component is open
 *    meetings
 *      Array of all meetings returned from the getAllMeetings query
 *    showPastMeetings
 *      Boolean state toggle indicating if past meetings are shown
 */

function MeetingListView() {
  const { t } = useTranslation();

  const { loading, error, data, refetch } = useQuery(GET_ALL_UPCOMING_MEETINGS);

  const [navToggled, setNavToggled] = useState(false);
  const [meetings, setMeetings] = useState([]);
  const [showPastMeetings, setShowPastMeetings] = useState(false);
  const isCurrentUserAdmin = isAdmin();

  function handleToggle() {
    setNavToggled(!navToggled);
  }

  useEffect(() => {
    if (data) {
      setMeetings(data.getAllUpcomingMeetings);
    }
    // poll for any meeting or agenda item status changes by other (admin) users
    const timer = window.setInterval(() => {
      refetch();
    }, PollIntervals.HALF_MINUTE_IN_MILLISECONDS);
    // clear interval polling timer when unmounting (e.g. user leaves page)
    return () => {
      clearInterval(timer);
    };
  }, [data]);

  const meetingsToDisplay = showPastMeetings
    ? meetings
    : meetings?.filter(
        (m) =>
          m.status === 'IN PROGRESS' ||
          isFutureTimestamp(m.meeting_start_timestamp)
      );
  const meetingGroups = groupMeetingsByDate(meetingsToDisplay);

  return (
    <div className="MeetingListView">
      <NavBarHeader toggled={navToggled} handleToggle={handleToggle} />

      <div className="meeting-list-header">
        <h2>{t('header.city-council-meetings')}</h2>
        {isCurrentUserAdmin ? <AdminMeetingListViewLinks /> : null}
      </div>

      <div className="meeting-list-content">
        {!(loading || error) &&
          (meetingGroups.length > 0 ? (
            <Accordion
              allowZeroExpanded
              allowMultipleExpanded
              preExpanded={[0]}
            >
              {meetingGroups.map((m, i) => (
                <MeetingListGroup
                  uuid={i}
                  key={`${m.month}${m.year}`}
                  month={m.month}
                  year={m.year}
                  meetings={m.meetings}
                />
              ))}
            </Accordion>
          ) : (
            <div>{t('meeting.list.no-meeting')}</div>
          ))}

        {loading && (
          <div className="loader">
            <Spinner />
          </div>
        )}
        {error && <div className="loader">{`Error! ${error.message}`}</div>}
        <div className="past-meetings">
          <p>{t('meeting.list.look-for-past')}</p>
          <a
            href={MEETINGS_ARCHIVE_URL.LINK}
            target="_blank" // open link in new tab
            rel="noopener noreferrer"
          >
            <p>
              Search the archive <ExternalSite className="translate-down" />
            </p>
          </a>
        </div>
        {/* <button
          type="button"
          className="complete-toggle"
          onClick={() => setShowPastMeetings((completed) => !completed)}
        >
          {showPastMeetings ? <CheckedCheckboxIcon /> : <UncheckedCheckboxIcon />}
          <p>{t('meeting.list.show-past')}</p>
        </button> */}
      </div>
    </div>
  );
}

export default MeetingListView;
