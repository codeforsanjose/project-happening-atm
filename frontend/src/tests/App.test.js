import React from 'react';
// npm install react-test-renderer version 17.2.0 for compatibility: https://stackoverflow.com/questions/72088446/uncaught-typeerror-cannot-read-properties-of-undefined-reading-isbatchinglega
import renderer from 'react-test-renderer';
import { AdminMeetingItemLinks } from '../components/MeetingListView/MeetingListItemLinks';
import { MockedProvider } from '@apollo/client/testing'; // Enable mocking of client: https://github.com/apollographql/apollo-client/issues/2042

// Mock to allow react-i18next in AdminMeetingItemLinks: https://stackoverflow.com/questions/66973857/next-i18next-jest-testing-with-usetranslation
jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}));

// Mock to allow i18next in 'frontend/src/utils/timestampHelper.js': https://stackoverflow.com/questions/70850517/testing-i18next-complain-about-cannot-read-property-use-of-undefined
jest.mock('i18next', () => ({
  __esModule: true,
  use: () => {},
  init: () => {},
  default: {
    t: (key) => key,
  },
}));

// Mock to allow useHistory: https://stackoverflow.com/questions/58392815/how-to-mock-usehistory-hook-in-jest
// Mock of <Link> as used in AdminMeetingItemLinks: https://stackoverflow.com/questions/64213791/react-testing-library-and-link-element-type-is-invalid-expected-a-string-or-a
jest.mock('react-router-dom', () => ({
  useHistory: jest.fn().mockReturnValue({
    location: { pathname: '/meeting/1' }, // pathname for view meeting #1 page
  }),
  Link: jest.fn().mockImplementation(({ children }) => {
    return children;
  }),
}));

// future effort: mock dayjs as used by timestampHelper.isFutureTimeStamp to capture proper classNames
it('renders admin mtg links on view mtg page that matches snapshot', () => {
  const tree = renderer
    .create(
      <MockedProvider>
        <AdminMeetingItemLinks meeting="__typename: 'meeting', id: 1, meeting_start_timestamp: '1703887201000', status: null" />
      </MockedProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Mock to allow useHistory: https://stackoverflow.com/questions/58392815/how-to-mock-usehistory-hook-in-jest
// Mock of <Link> as used in AdminMeetingItemLinks: https://stackoverflow.com/questions/64213791/react-testing-library-and-link-element-type-is-invalid-expected-a-string-or-a
jest.mock('react-router-dom', () => ({
  useHistory: jest.fn().mockReturnValue({
    location: { pathname: '/' }, // pathname of main view meetings dashboard page
  }),
  Link: jest.fn().mockImplementation(({ children }) => {
    return children;
  }),
}));

// future effort: mock dayjs as used by timestampHelper.isFutureTimeStamp to capture proper classNames
it('renders admin mtg links on view mtgs dashboard page that matches snapshot', () => {
  const tree = renderer
    .create(
      <MockedProvider>
        <AdminMeetingItemLinks meeting="__typename: 'meeting', id: 1, meeting_start_timestamp: '1703887201000', status: null" />
      </MockedProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
