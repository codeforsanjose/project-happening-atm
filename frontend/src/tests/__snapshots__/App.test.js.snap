// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders admin mtg links on view mtg page that matches snapshot 1`] = `
<div
  className="meeting-links visibility-hidden"
>
  <button
    className="link upload-agenda-button"
    onClick={[Function]}
    type="button"
  >
    <svg>
      view_agenda-24px.svg
    </svg>
    <p>
      meeting.actions.upload-new-agenda
    </p>
  </button>
  <button
    className="link delete-agenda-button"
    onClick={[Function]}
    type="button"
  >
    <svg>
      delete-24px.svg
    </svg>
    <p>
      meeting.list.delete-meeting.button
    </p>
  </button>
</div>
`;

exports[`renders admin mtg links on view mtgs dashboard page that matches snapshot 1`] = `
<div
  className="meeting-links visibility-hidden"
>
  <button
    className="link upload-agenda-button"
    onClick={[Function]}
    type="button"
  >
    <svg>
      view_agenda-24px.svg
    </svg>
    <p>
      meeting.actions.upload-new-agenda
    </p>
  </button>
  <button
    className="link delete-agenda-button"
    onClick={[Function]}
    type="button"
  >
    <svg>
      delete-24px.svg
    </svg>
    <p>
      meeting.list.delete-meeting.button
    </p>
  </button>
</div>
`;
